### PR TITLE
feat: switch to metrics_derive

### PR DESCRIPTION
**Pull request by andysim3d**
[Closes/Fixes] #166

## Proposed Changes

  -
  -
  -

### CODE REVIEW

File: bin/rundler/Cargo.toml [hunk 1]
```diff
@@ -47,3 +47,4 @@ tracing.workspace = true
 tracing-appender = "0.2.3"
 tracing-log = "0.2.0"
 tracing-subscriber = { version = "0.3.18", features = ["env-filter", "fmt", "json"] }
+metrics-derive = "0.1.0"
```
Comment (dancoombs) [inline bin/rundler/Cargo.toml:50]: On all of these you should put this in the top level `Cargo.toml` and then use `metrics-derive.workspace` in all of the crates/bin

File: bin/rundler/src/cli/metrics.rs [hunk 2]
```diff
@@ -79,77 +82,202 @@ pub fn initialize<'a, T: TaskSpawner>(
     Ok(())
 }
 
-const TOKIO_PREFIX: &str = "tokio_rt_";
+#[allow(dead_code)]
+#[derive(Metrics)]
+#[metrics(scope = "rundler_tokio_rt")]
```
Comment (dancoombs) [inline bin/rundler/src/cli/metrics.rs:87]: Does `rundler_` not already get added at the top level (line 52/53)? Or does this do something different?

File: crates/builder/src/bundle_sender.rs [hunk 3]
```diff
@@ -154,7 +156,7 @@ where
         loop {
             if let Err(e) = self.step_state(&mut state).await {
                 error!("Error in bundle sender loop: {e:#?}");
-                self.metrics.increment_state_machine_errors();
+                self.metrics.state_machine_errors.increment(1_u64);
```
Comment (dancoombs) [inline crates/builder/src/bundle_sender.rs:159]: nit: do you need the `_u64` on all of these?

File: crates/builder/src/bundle_sender.rs [hunk 4]
```diff
@@ -1127,90 +1132,60 @@ impl BundleSenderTrigger {
     }
 }
 
-#[derive(Debug, Clone)]
-struct BuilderMetrics {
-    builder_index: u64,
-    entry_point: Address,
+#[derive(Metrics)]
+#[metrics(scope = "builder")]
+struct BuilderMetric {
```
Comment (dancoombs) [inline crates/builder/src/bundle_sender.rs:1137]: This is SO much cleaner. Love it.

File: crates/builder/src/signer/mod.rs [hunk 5]
```diff
@@ -63,14 +65,21 @@ pub(crate) trait Signer: Send + Sync {
...
+
 pub(crate) async fn monitor_account_balance<P: EvmProvider>(addr: Address, provider: P) {
     loop {
         match provider.get_balance(addr, None).await {
             Ok(balance) => {
                 let eth_balance = balance.to_f64().unwrap_or_default() / 1e18;
                 tracing::info!("account {addr:?} balance: {}", eth_balance);
-                metrics::gauge!("bundle_builder_account_balance", "addr" => format!("{addr:?}"))
-                    .set(eth_balance);
+                let metric = BuilderMetric::new_with_labels(&[("addr", format!("{addr:?}"))]);
```
Comment (dancoombs) [inline crates/builder/src/signer/mod.rs:81]: Can we do this once before the loop?

File: crates/builder/src/transaction_tracker.rs [hunk 6]
```diff
@@ -474,31 +490,19 @@ impl From<TxSenderError> for TransactionTrackerError {
...
-            .set(fees.max_fee_per_gas as f64);
-        metrics::gauge!("builder_tracker_current_max_priority_fee_per_gas", "builder_index" => builder_index.to_string())
-            .set(fees.max_priority_fee_per_gas as f64);
-    }
+#[derive(Metrics)]
+#[metrics(scope = "builder_tracker_")]
+struct TransactionTrackerMetrics {
+    #[metric(describe = "the number of pending transactions.")]
+    num_pending_transactions: Gauge,
+    #[metric(describe = "the nonce of transaction.")]
```
Comment (dancoombs) [inline crates/builder/src/transaction_tracker.rs:498]: nit: this should say "the current account nonce"

File: crates/builder/src/bundle_sender.rs [hunk 7]
```diff
@@ -1127,90 +1132,60 @@ impl BundleSenderTrigger {
...
+    #[metric(describe = "the number of bundle transactions already sent.")]
```
Comment (dancoombs) [inline crates/builder/src/bundle_sender.rs:1138]: nit: we use "the number of ..." in places where we use Counters and Guages. Maybe we differentiate between the two.

For counters -> "count of"
For gauges -> "current number of"

File: crates/builder/src/transaction_tracker.rs [hunk 8]
```diff
@@ -474,31 +490,19 @@ impl From<TxSenderError> for TransactionTrackerError {
...
+    nonce: Gauge,
+    #[metric(describe = "the number of pending transactions.")]
+    attempt_count: Gauge,
+    #[metric(describe = "the maximum fee per gas of current transaction.")]
+    current_max_fee_per_gas: Gauge,
+    #[metric(describe = "the maximum priority fee per gas of current transaction.")]
+    max_priority_fee_per_gas: Gauge,
```
Comment (dancoombs) [inline crates/builder/src/transaction_tracker.rs:505]: Looks like this drops the `current_` from the already existing metric. We should be consistent with `max_fee_per_gas`. I'm happy to drop the `current_` (its not needed), but we will need to remember to update dashboards.

File: crates/pool/src/chain.rs [hunk 9]
```diff
@@ -698,28 +703,20 @@ impl ChainUpdate {
...
-    fn increment_sync_retries() {
-        metrics::counter!("op_pool_chain_sync_retries").increment(1);
-    }
-
-    fn increment_sync_abandoned() {
-        metrics::counter!("op_pool_chain_sync_abandoned").increment(1);
-    }
+#[derive(Metrics)]
+#[metrics(scope = "op_pool_chain")]
+
```
Comment (dancoombs) [inline crates/pool/src/chain.rs:708]: nit: remove newline

File: crates/pool/src/chain.rs [hunk 10]
```diff
@@ -698,28 +703,20 @@ impl ChainUpdate {
...
+struct ChainMetrics {
+    #[metric(describe = "the height of block.")]
+    block_height: Gauge,
+    #[metric(describe = "the number of reorg event detected.")]
+    reorgs_detected: Counter,
+    #[metric(describe = "the total number of reorg depth.")]
+    total_reorg_depth: Counter,
+    #[metric(describe = "the number of removed entities.")]
```
Comment (dancoombs) [inline crates/pool/src/chain.rs:716]: nit: wrong description

File: crates/pool/src/mempool/pool.rs [hunk 11]
```diff
@@ -600,41 +608,23 @@ impl TimeToMineInfo {
...
-        .record(time_to_mine.candidate_for_blocks as f64);
-    }
+#[derive(Metrics)]
+#[metrics(scope = "op_pool")]
+struct PoolMetrics {
+    #[metric(describe = "the number of ops in mempool.")]
+    num_ops_in_pool: Gauge,
+    #[metric(describe = "the size of mempool in bytes.")]
+    size_bytes: Gauge,
+    #[metric(describe = "the number of ops in mempool cache(mined but not persistent).")]
```
Comment (dancoombs) [inline crates/pool/src/mempool/pool.rs:618]: nit: add space after `cache`

File: crates/pool/src/mempool/pool.rs [hunk 12]
```diff
@@ -600,41 +608,23 @@ impl TimeToMineInfo {
...
+    num_ops_in_cache: Gauge,
+    #[metric(describe = "the size of mempool cache in bytes.")]
+    cache_size_bytes: Gauge,
+    #[metric(describe = "the number of candidates.")]
+    num_candidates: Gauge,
+    #[metric(describe = "the duration of a bundle mined.")]
```
Comment (dancoombs) [inline crates/pool/src/mempool/pool.rs:624]: nit: we should describe this as a histogram of user operation's times to mine

File: crates/pool/src/mempool/pool.rs [hunk 13]
```diff
@@ -600,41 +608,23 @@ impl TimeToMineInfo {
...
+    time_to_mine: Histogram,
+    #[metric(describe = "the duration of a blcoked mined.")]
```
Comment (dancoombs) [inline crates/pool/src/mempool/pool.rs:626]: nit: same as above, and typo

File: crates/pool/src/mempool/uo_pool.rs [hunk 14]
```diff
@@ -691,40 +707,28 @@ where
...
+#[derive(Metrics)]
+#[metrics(scope = "op_pool")]
+struct UoPoolMetrics {
+    #[metric(describe = "the maximum fee in Gwei.")]
+    current_max_fee_gwei: Gauge,
+    #[metric(describe = "the maximum priority fee in Gwei.")]
+    current_max_priority_fee_gwei: Gauge,
+    #[metric(describe = "the base fee of current block.")]
+    current_base_fee: Gauge,
 }
```
Comment (dancoombs) [inline crates/pool/src/mempool/uo_pool.rs:732]: What do you think about simplifying this into a single struct and just handling the duplication of the fees in dashboards?

File: crates/types/src/task/metric_recorder.rs [hunk 15]
```diff
@@ -22,93 +23,83 @@ pub struct MethodSessionLogger {
     service_name: String,
     method_name: String,
     protocol: String,
+    method_metric: MethodMetrics,
+}
+
+#[derive(Metrics)]
+#[metrics(scope = "rundler_runtime")]
```
Comment (dancoombs) [inline crates/types/src/task/metric_recorder.rs:30]: Do you need `rundler_`

Comment (dancoombs) [reply crates/types/src/task/metric_recorder.rs:30]: Also is this the correct scope?

File: crates/types/src/task/metric_recorder.rs [hunk 16]
```diff
@@ -22,93 +23,83 @@ pub struct MethodSessionLogger {
...
+pub(crate) struct MethodMetrics {
+    #[metric(describe = "total num of requests.")]
+    num_requests: Counter,
+
+    #[metric(describe = "opening requests.")]
```
Comment (dancoombs) [inline crates/types/src/task/metric_recorder.rs:35]: nit: open requests

File: crates/types/src/task/metric_recorder.rs [hunk 17]
```diff
@@ -22,93 +23,83 @@ pub struct MethodSessionLogger {
...
+    open_requests: Gauge,
+
+    #[metric(describe = "opening requests.")]
```
Comment (dancoombs) [inline crates/types/src/task/metric_recorder.rs:38]: nit: wrong description

File: crates/types/src/task/metric_recorder.rs [hunk 18]
```diff
@@ -22,93 +23,83 @@ pub struct MethodSessionLogger {
...
+    request_latency: Histogram,
+}
+
+#[derive(Metrics)]
+#[metrics(scope = "rundler_runtime")]
+pub(crate) struct MethodStatusMetrics {
+    #[metric(describe = "opening requests.")]
```
Comment (dancoombs) [inline crates/types/src/task/metric_recorder.rs:45]: Wrong description

File: crates/types/src/task/metric_recorder.rs [hunk 19]
```diff
@@ -22,93 +23,83 @@ pub struct MethodSessionLogger {
...
+    http_response_status: Counter,
+
+    #[metric(describe = "opening requests.")]
```
Comment (dancoombs) [inline crates/types/src/task/metric_recorder.rs:48]: Wrong description

File: crates/types/src/task/metric_recorder.rs [hunk 20]
```diff
@@ -22,93 +23,83 @@ pub struct MethodSessionLogger {
...
     /// record a rpc status code.
     pub fn record_rpc(&self, rpc_code: RpcCode) {
-        MethodMetrics::increment_rpc_response_code(&self.method_name, &self.service_name, rpc_code);
+        let method_status_metric = MethodStatusMetrics::new_with_labels(&[
+            ("method_name", self.method_name.clone()),
+            ("service_name", self.service_name.clone()),
+            ("protocol", self.protocol.clone()),
+            ("status_code", rpc_code.to_string()),
+        ]);
+        method_status_metric.rpc_response_status.increment(1);
```
Comment (dancoombs) [inline crates/types/src/task/metric_recorder.rs:84]:
```suggestion
        MethodStatusMetrics::new_with_labels(&[
            ("method_name", self.method_name.clone()),
            ("service_name", self.service_name.clone()),
            ("protocol", self.protocol.clone()),
            ("status_code", rpc_code.to_string()),
        ]).increment(1);
```

File: crates/builder/src/bundle_sender.rs [hunk 21]
```diff
@@ -1127,90 +1132,60 @@ impl BundleSenderTrigger {
...
+    bundle_gas_limit: Counter,
+    #[metric(describe = "the number of bundle gas used.")]
+    bundle_gas_used: Counter,
+    #[metric(describe = "the number of dropped bundle transactions.")]
+    bundle_txns_dropped: Counter,
+    #[metric(describe = "the number of anabdoned bundle transactions.")]
+    bundle_txns_abandoned: Counter,
+    #[metric(describe = "the number of failed bundle transactions.")]
+    bundle_txns_failed: Counter,
+    #[metric(describe = "the number of bundle transactin nonce used event.")]
```
Comment (dancoombs) [inline crates/builder/src/bundle_sender.rs:1152]: nit: typo transactin -> transaction, event -> events (same below)

File: crates/pool/src/mempool/uo_pool.rs (hunk 14 above)
Comment (andysim3d) [reply crates/pool/src/mempool/uo_pool.rs:732]: that might confuse who don't familiar the metric setup. they might find the value doubled (for 2 EPs we supportted.)

**Review by dancoombs — APPROVED** [review]
🚢

**Pull request merged**